Contest_API.md

```diff
@@ -230,6 +230,10 @@ absolute timestamps.
   containing human-readable time durations, given in a slight modification of
   the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) extended time format:
   `(-)?(h)*h:mm:ss(.uuu)?`
+- Time intervals (type **`INTERVAL`** in the specification) are strings 
```

Comment:
What if the end of the interval is not known yet.

I think that is a case @johnbrvc and @clevengr would like to have, where they 'pause' the contest and so only know the start of the interval and later update it. Or are you saying in that case they should only emit a state update event once the end is also known?

Reply:
Agreed, one of the use cases is something like a fire alarm where you literally hit the 'stop' button and walk out the door. Yes, we could leave the clock running in subsystems and set the end/update the event afterward, but it would be nice for end time to be optional.

Reply:
Do we *want* the spec to allow for open intervals? I'm not in favor or against, but if we are thinking about implementing removed intervals properly now, this is the time to discuss this I'd say.

Reply:
I see Fredrik has already 👎🏼 our comments and maybe we don't. 🙂 I'm just not sure I love forcing one which could end up with either nothing sent ('contest is still running' when clearly it isn't) or far future ('contest will resume in 1865 days') is great either.

Instead of adding a new type (will we need INTERVAL anywhere else?) why not a child object like location: `{ start: TIME, end: TIME }`?

Reply:
> Do we _want_ the spec to allow for open intervals? I'm not in favor or against, but if we are thinking about implementing removed intervals properly now, this is the time to discuss this I'd say.

I don't think we do. The simplest argument is that there is always a trivial workaround. Just pick any large enough end and you practically have an open interval anyway.

> I'm just not sure I love forcing one which could end up with either nothing sent ('contest is still running' when clearly it isn't) 

Well, it **is** though. If it actually isn't running then it has ended, and there are other messages you should be sending.

> or far future ('contest will resume in 1865 days') is great either.

Yeah, I agree that this is a bit ugly, but it's or the most cornery of cases, I think it's fine.

> Instead of adding a new type (will we need INTERVAL anywhere else?) why not a child object like location: `{ start: TIME, end: TIME }`?

That's a great point, and in fact that is exactly how I first wrote it. That said, there is a standard ISO format for time intervals in ISO 8601, and we use ISO 8601 formats whenever available elsewhere, so it makes sense to do the same here. I think as I wrote it now is cleaner and better, but I don't feel *very* strongly about that.

Reply:
> > I'm just not sure I love forcing one which could end up with either nothing sent ('contest is still running' when clearly it isn't)
> 
> Well, it **is** though. If it actually isn't running then it has ended, and there are other messages you should be sending.

I meant more that if all the CCS has is a pause and resume capability, they might decide not to send an event on pause, which would be missing the point of this change. We should set the expectation that it should send an event on pause with a long interval, and make sure someone from PC^2 has reviewed this too b/c I think that's what they have.

> > Instead of adding a new type (will we need INTERVAL anywhere else?) why not a child object like location: `{ start: TIME, end: TIME }`?
> 
> That's a great point, and in fact that is exactly how I first wrote it. That said, there is a standard ISO format for time intervals in ISO 8601, and we use ISO 8601 formats whenever available elsewhere, so it makes sense to do the same here. I think as I wrote it now is cleaner and better, but I don't feel _very_ strongly about that.

I lean slightly the other way that it's not worth adding a global type for this when time would be enough - but I don't feel very strongly about it either. :)

Reply:
> I meant more that if all the CCS has is a pause and resume capability, they might decide not to send an event on pause, which would be missing the point of this change. 

Sure, but they also could send an event. 

> We should set the expectation that it should send an event on pause with a long interval,

Do you mean adding a comment to the effect of "If the end of the removed interval is not known, a system should provide a best estimate, or at least some very large (such as 1 day later) value for the end"?

Reply:
I don't like having a removed interval with an end time far in the future just to "represent" the fact that it is unknown. That means that the interval needs to be updated again when the end time becomes known, and that can lead to extra inconsistencies. What if the "far in the future" end time actually passes without anyone paying attention?

I'm not really a fan of supporting open ended intervals, but I strongly feel that if we do, we should do so properly by having separate start and end fields and that the end can be `null`.

Furthermore, ISO 8601 supports three different alternative formats for well-defined intervals (not just durations):
```
<start>/<end>
<start>/<duration>
<duration>/<end>
```
I don't really see the benefit of allowing all that over simply specifying a start and end date/time in normal ISO 8601 format.

Reply:
[Somehow it seemed I added a duplicate comment yesterday, then removed one, and both were gone. Hree goes again.]

I think I'm not too much a fan of open-ended removed intervals, but if we support that, I object to using an end time far in the future to indicate that. I think it's misleading and might lead to problems, for example, if that end time is not that far in the future and it passes without someone noticing. Also, it means that this end time will get modified again, and I think we should try to write the specification to make things as immutable as possible.

Also, in general, I feel like it would be better to just use separate ISO date/time fields for start and end time, because a (well-defined) interval can have various formats:
```
<start>/<end>
<start>/<duration>
<duration>/<end>
```
I don't see the benefit of allowing these.